partiql-parser/src/parse/mod.rs

```diff
@@ -32,42 +36,64 @@ type LalrpopResult<'input> = Result<Box<ast::Expr>, LalrpopError<'input>>;
 type LalrpopErrorRecovery<'input> =
     lpop::ErrorRecovery<ByteOffset, lexer::Token<'input>, ParseError<'input, BytePosition>>;
 
-pub(crate) type AstResult<'input> = Result<Box<ast::Expr>, Vec<ParseError<'input, BytePosition>>>;
+#[derive(Debug, Clone)]
+pub(crate) struct AstData {
+    pub ast: Box<ast::Expr>,
+    pub locations: LocationMap<NodeId>,
+    pub offsets: LineOffsetTracker,
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct ErrorData<'input> {
+    pub errors: Vec<ParseError<'input, BytePosition>>,
+    pub offsets: LineOffsetTracker,
+}
+
+pub(crate) type AstResult<'input> = Result<AstData, ErrorData<'input>>;
 
 lazy_static! {
     static ref BUILT_INS: FnExprSet<'static> = built_ins();
 }
 
 /// Parse PartiQL query text.
-pub fn parse_partiql<'input>(s: &'input str, offsets: &mut LineOffsetTracker) -> AstResult<'input> {
-    let mut errors: Vec<LalrpopErrorRecovery> = vec![];
-    let lexer = PreprocessingPartiqlLexer::new(s, offsets, &*BUILT_INS);
-
-    let parsed: LalrpopResult = grammar::QueryParser::new().parse(s, &mut errors, lexer);
-
-    process_errors(s, offsets, parsed, errors)
+pub(crate) fn parse_partiql(s: &str) -> AstResult {
+    parse_partiql_with_state(s, ParserState::default())
 }
 
-fn process_errors<'input>(
-    _s: &'input str,
-    _offsets: &LineOffsetTracker,
-    result: Result<Box<ast::Expr>, LalrpopError<'input>>,
-    errors: Vec<LalrpopErrorRecovery<'input>>,
+pub(crate) fn parse_partiql_with_state<'input, Id: IdGenerator>(
```

Comment:
Is it intentional for this to be public? If so, may need some documentation for the function.

Reply:
Despite the name `pub(crate)` really means "crate-private" and is thus visible to only other functions in the current crate.

Reply:
But, also, I think it can be module private; will update.

Reply:
> Despite the name `pub(crate)` really means "crate-private" and is thus visible to only other functions in the current crate.

Ah, my mistake. Mixed up the API visibilities with other languages. Thanks for the reference;

Reply:
Yeah. Rust's API visibility is somewhat expressive, but can be confusing, as the effects are local, and thus you need a global understanding of the crate's structure to understand the public exports.